File: android/src/main/java/ch/nevis/mobile/authentication/sdk/react/userinteraction/selector/AccountSelectorImpl.kt
/*
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

package ch.nevis.mobile.authentication.sdk.react.userinteraction.selector

import ch.nevis.mobile.authentication.sdk.react.cache.OperationCache
import ch.nevis.mobile.authentication.sdk.react.cache.operation.UserInteractionOperation
import ch.nevis.mobile.authentication.sdk.react.cache.state.SelectAccount
import ch.nevis.mobile.authentication.sdk.react.emitter.EventEmitter
import ch.nevis.mobile.authentication.sdk.react.model.ReactMethodName
import ch.nevis.mobile.authentication.sdk.react.model.message.outgoing.AccountSelectionMessage
import ch.nevis.mobile.authentication.sdk.react.model.sdk.toDto
import ch.nevis.mobile.sdk.api.operation.selection.AccountSelectionContext
import ch.nevis.mobile.sdk.api.operation.selection.AccountSelectionHandler
import ch.nevis.mobile.sdk.api.operation.selection.AccountSelector

class AccountSelectorImpl(
    private val operationId: String
) : AccountSelector {

    override fun selectAccount(
        accountSelectionContext: AccountSelectionContext?,
        accountSelectionHandler: AccountSelectionHandler?
    ) {
        val operation = OperationCache.getOperation<UserInteractionOperation>(operationId)
        OperationCache.updateOperation(
            operationId,
            operation.update(SelectAccount(accountSelectionContext, accountSelectionHandler))
        )

        accountSelectionContext?.also {
            val message = AccountSelectionMessage(operationId, it.toDto(operationId))
            EventEmitter.dispatch(ReactMethodName.SELECT_ACCOUNT, message)
        } ?: throw IllegalStateException("AccountSelectionContext cannot be null.")
    }
}

File: ios/UserInteraction/Selector/AccountSelectorImpl.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

class AccountSelectorImpl {
	// MARK: Properties

	let operationId: String

	// MARK: Initialization

	init(operationId: String) {
		self.operationId = operationId
	}
}

// MARK: - AccountSelector

extension AccountSelectorImpl: AccountSelector {
	func selectAccount(context: AccountSelectionContext, handler: AccountSelectionHandler) {
		do {
			let operation: UserInteractionOperation = try OperationCache.shared.read(by: operationId)
			let state = SelectAccountState(context: context, handler: handler)
			OperationCache.shared.update(by: operationId,
			                             operation: operation.update(state: state))

			let message = AccountSelectorMessage(operationId: operationId,
			                                     context: context)
			EventEmitter.shared.dispatch(event: .selectAccount, message: message)
		}
		catch {
			OperationCache.shared.delete(by: operationId)
			fatalError("Failed to select account! Error: \(error.localizedDescription)")
		}
	}
}
